### PR TITLE
Fix an failed assertion in HttpSM::parse_range_and_compare

### PR DESCRIPTION
**Pull request by skyever**
fix the issue #3007 

when using higher version gcc to build the ats with optimization options such as O2 or O3, the gcc will think  the start value always be positive and ignore the conditions if start < 0. So when client sends request with a large range number such as 18446744073709551614-10000,  the start value is negative, so the assert 
```
 4315   ink_assert(start >= 0 && end >= 0 && start < content_length && end < content_length);
```
will fail. 

Signed-off-by: fengshuaitao <fengshuaitao@bytedance.com>

### CODE REVIEW

File: proxy/http/HttpSM.cc [hunk 1]
```diff
@@ -4186,6 +4186,8 @@ HttpSM::parse_range_and_compare(MIMEField *field, int64_t content_length)
   const char *s, *e, *tmp;
   RangeRecord *ranges = nullptr;
   int64_t start, end;
+  int64_t cutoff = 9223372036854775807 / 10;
+  int64_t cutlim = 9223372036854775807 % 10;
```
Comment (shinrich) [inline proxy/http/HttpSM.cc]: Specifying the constant as 0x7FFFFFFFFFFFFFFF might be more meaningful.   Or better yet, perhaps a named constant.

Comment (shinrich) [reply proxy/http/HttpSM.cc]: However, the divide and modulo 10 seem kind of arbitrary.  Are we just trying to work around a compiler bug?

**Comment by shinrich** [comment]
[approve ci]

File: proxy/http/HttpSM.cc (hunk 1 above)
Comment (zwoop) [reply proxy/http/HttpSM.cc]: Or even a INT64_MAX might be nicer here ?

**Comment by maskit** [comment]
> when using higher version gcc to build the ats with optimization options such as O2 or O3, the gcc will think the start value always be positive and ignore the conditions if start < 0.

This makes sense. I extracted the function and did brute-force with all possible (well-formed) header values but I couldn't find any combinations that fail the assert.

File: proxy/http/HttpSM.cc (hunk 1 above)
Comment (skyever) [reply proxy/http/HttpSM.cc]: INT64_MAX is a better choice.

**Comment by maskit** [comment]
@skyever I used clang on macOS. I'll try out your test code on the same machine later.

**Comment by maskit** [comment]
It seems that gcc 7.1.1 does the same problematic optimization.

on macOS
```
[masakazu@MacBook:tmp]$ clang -O test.c 
[masakazu@MacBook:tmp]$ ./a.out 
start: -2 < 0
[masakazu@MacBook:tmp]$ clang -O3 test.c 
[masakazu@MacBook:tmp]$ ./a.out 
start: -2 < 0
[masakazu@MacBook:tmp]$ clang -v
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.3.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

on Fedora26
```
[maskit@localhost ~]$ gcc test.c  -O
[maskit@localhost ~]$ ./a.out 
start: -2 < 0
[maskit@localhost ~]$ gcc test.c  -O3
[maskit@localhost ~]$ ./a.out 
start: -2 >= 0
[maskit@localhost ~]$ gcc -v
Using built-in specs.
COLLECT_GCC=/usr/bin/gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/7/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,objc,obj-c++,fortran,ada,go,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl --enable-libmpx --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
gcc version 7.1.1 20170622 (Red Hat 7.1.1-3) (GCC) 
```

**Comment by maskit** [comment]
[approve ci]

**Comment by shinrich** [comment]
Comments in the code that this works around a compiler optimization flaw would be good.  Otherwise, this will cause head scratching in years to come.

**Comment by scw00** [comment]
[approve ci]

**Comment by maskit** [comment]
@skyever Could you run `make clang-format` and squash the commits?

**Comment by maskit** [comment]
[approve ci]

File: proxy/http/HttpSM.cc [hunk 2]
```diff
@@ -4186,6 +4186,8 @@ HttpSM::parse_range_and_compare(MIMEField *field, int64_t content_length)
   const char *s, *e, *tmp;
   RangeRecord *ranges = nullptr;
   int64_t start, end;
+  int64_t cutoff = INT64_MAX / 10;
```
Comment (zwoop) [inline proxy/http/HttpSM.cc]: This ought to be const static, or constexpr or some such.

File: proxy/http/HttpSM.cc [hunk 3]
```diff
@@ -4242,6 +4244,12 @@ HttpSM::parse_range_and_compare(MIMEField *field, int64_t content_length)
       start = -1;
     } else {
       for (start = 0; s < e && *s >= '0' && *s <= '9'; ++s) {
+        // check the int64 overflow in case of high gcc with O3 option
+        // thinking the start is always positive
+        if (start >= cutoff && (start > cutoff || *s - '0' > cutlim)) {
```
Comment (zwoop) [inline proxy/http/HttpSM.cc]: To the naked eye, this looks really weird. The whole checking of start twice, with slightly different range. But maybe this is needed as is ?

**Comment by bryancall** [comment]
[approve ci debian]

**Comment by bryancall** [comment]
@skyever Can you please update the PR based on the comments?

**Comment by zwoop** [comment]
Status on this?

**Comment by zwoop** [comment]
Ping?

File: proxy/http/HttpSM.cc (hunk 3 above)
Comment (zwoop) [reply proxy/http/HttpSM.cc]: I'm thinking the if conditional reads better with e.g.

```
(start > cutoff) || (start == cutoff && (*s -'0' > cutlim))
```

Comment (zwoop) [reply proxy/http/HttpSM.cc]: And Maybe a comment like:

```
// Start > cutoff,  or when at cutoff and last digit > cutlim (8), we would overflow the additions to start
```


Or, after discussion with Bryan, maybe we should just use strtoll() here, which deals with overflow as well.

**Comment by zwoop** [comment]
or maybe do

```
int64_t new = start * 10 + (*s - '0');

if (new < start) {
 // failure case ...
} else {
    start = new;
}
```

or some such, whatever way we'd detect that new overflowed :).

**Pull request merged**